src/krotov/propagators.py

```diff
@@ -266,11 +280,17 @@ def _rhs(t, rho, L_list):
         out = np.zeros(rho.shape[0], dtype=complex)
         L = L_list[0][0]
         L_coeff = L_list[0][1]
-        spmvpy_csr(L.data, L.indices, L.indptr, rho, L_coeff, out)
+        if is_qutip5:
+            out = L_coeff * L @ rho
+        else:
+            spmvpy_csr(L.data, L.indices, L.indptr, rho, L_coeff, out)
         for n in range(1, len(L_list)):
             L = L_list[n][0]
             L_coeff = L_list[n][1]
-            spmvpy_csr(L.data, L.indices, L.indptr, rho, L_coeff, out)
+            if is_qutip5:
+                out = L_coeff * L @ rho
```

Comment:
`spmvpy_csr` has an implicit `out +=` that was forgotten.

```suggestion
                out = qutip.data.add(qutip.data.matmul(L, rho, scale=L_coeff), out)
```
We have `imatmul_data_dense(left, right, scale, out)` that does the matmul implace like `spmvpy_csr`, but no python binding. (It's only used in `QobjEvo`.)

Reply:
If I keep things in `qutip.Qobj`, not NumPy array, since we have a much better initialization of `Qobj` now in v5. Will `sparse_csr_qobj * dense_ket_qobj` automatically use that `imatmul_data_dense`.

This will allow the switch between dense and csr.

Reply:
Since the states, (out, rho) are used by scipy's ODE solver they must enter and exit as numpy's array.
`imatmul_data_dense` is equivalent to `out += left @ right * scale`, this can't be represented in `Qobj.__matmul__` operation so no. `QobjEvo.matmul_data(t, state, out)` would use the inplace operation if out is given and dense. But this use the list format directly and rewriting to use `QobjEvo` is probably not a small task.

`qutip.data.add` and `qutip.data.matmul` use qutip's data that can be sparse, dense, jax etc. so it should enough to be able to switch between format.